### PR TITLE
[JENKINS-73474] Add `GitHubAppCredentials.withOwner` non regression test

### PR DESCRIPTION
**Pull request by jeromepochat**
A regression around `GitHubAppCredentials.withOwner` was introduced since https://github.com/jenkinsci/github-branch-source-plugin/pull/796.

As a follow-up to the [fix](https://github.com/jenkinsci/github-branch-source-plugin/pull/803), this adds non regression test.

### CODE REVIEW

**Review by jglick — APPROVED** [review]
A unit test, OK. What we would really like is an integration test with WireMock of running a multibranch pipeline and having the various involved plugins wire through the `Run` context that allows `withOwner` to be called with the right value. That would be a lot more work of course.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java [hunk 1]
```diff
@@ -56,40 +55,6 @@ public class GithubAppCredentialsTest extends AbstractGitHubWireMockTest {
     private static GitHubAppCredentials appCredentials, appCredentialsNoOwner;
     private static LogRecorder logRecorder;
 
-    // https://stackoverflow.com/a/22176759/4951015
-    public static final String PKCS8_PRIVATE_KEY = "-----BEGIN PRIVATE KEY-----\n"
```
Comment (jeromepochat) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsTest.java:60]: Extracted to `GitHubApp` to reuse from multiple test classes

**Comment by jeromepochat** [comment]
:eyes: why the test is failing on windows but pass on linux.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+import java.time.format.DateTimeFormatter;
+import java.time.temporal.ChronoUnit;
+import java.util.Arrays;
+import java.util.Date;
+import jenkins.branch.BranchSource;
+import org.jenkinsci.plugins.workflow.multibranch.WorkflowMultiBranchProject;
+import org.junit.Before;
+import org.junit.Test;
+
+public class GithubAppCredentialsContextualizationTest extends AbstractGitHubWireMockTest {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java:25]:
```suggestion
public class GitHubAppCredentialsContextualizationTest extends AbstractGitHubWireMockTest {
```
plus file rename

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java [hunk 3]
```diff
@@ -0,0 +1,99 @@
...
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBodyFile("../contextualization/body-repos-cloudbeers-multibranch-demo-contents.json")));
+        githubApi.stubFor(get(urlEqualTo("/repos/cloudbeers/multibranch-demo/pulls?state=open"))
+                .willReturn(aResponse()
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBodyFile("../contextualization/body-repos-cloudbeers-multibranch-demo-pulls.json")));
+    }
+
+    @Test
+    public void ownerMustBeInferedFromRepository() throws Exception {
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java:70]:
```suggestion
    public void ownerMustBeInferredFromRepository() throws Exception {
```

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+        credentials.setApiUri(githubApi.baseUrl());
+
+        final var scmSource = new GitHubSCMSource("cloudbeers", "multibranch-demo", null, false);
+        scmSource.setTraits(Arrays.asList(new BranchDiscoveryTrait(true, true)));
+        scmSource.setCredentialsId(credentials.getId());
+        scmSource.setApiUri(githubApi.baseUrl());
+
+        final var multiBranchProject = r.jenkins.createProject(WorkflowMultiBranchProject.class, "multibranch-demo");
+        multiBranchProject.getSourcesList().add(new BranchSource(scmSource));
+        multiBranchProject.scheduleBuild2(0, new Action[0]).getFuture().get();
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java:84]:
```suggestion
        multiBranchProject.scheduleBuild2(0).getFuture().get();
```
works I think

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java [hunk 5]
```diff
@@ -0,0 +1,99 @@
...
+        final var branchProject = multiBranchProject.getItem("master");
+        assertThat(branchProject, notNullValue());
+    }
+
+    private String createTokenExpiration() {
+        // This token will go stale at the soonest allowed time but will not
+        // expire for the duration of the test
+        // Format: 2019-08-10T05:54:58Z
+        return DateTimeFormatter.ISO_INSTANT.format(Instant.ofEpochMilli(new Date(System.currentTimeMillis()
+                                + Duration.ofMinutes(10).toMillis())
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java:95]: I think you can use `Instant.now().add(Duration.ofMinutes(10))` or similar, to be more idiomatic.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+        scmSource.setTraits(Arrays.asList(new BranchDiscoveryTrait(true, true)));
+        scmSource.setCredentialsId(credentials.getId());
+        scmSource.setApiUri(githubApi.baseUrl());
+
+        final var multiBranchProject = r.jenkins.createProject(WorkflowMultiBranchProject.class, "multibranch-demo");
+        multiBranchProject.getSourcesList().add(new BranchSource(scmSource));
+        multiBranchProject.scheduleBuild2(0, new Action[0]).getFuture().get();
+
+        final var branchProject = multiBranchProject.getItem("master");
+        assertThat(branchProject, notNullValue());
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java:87]: And without the fix, is this the line that fails?

**Review by jglick — APPROVED** [review]
Thanks!

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java (hunk 6 above)
Comment (jeromepochat) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java]: > And without the fix, is this the line that fails?

Yes. I confirm that the test is failing without https://github.com/jenkinsci/github-branch-source-plugin/pull/803

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java (hunk 2 above)
Comment (jeromepochat) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java]: https://github.com/jenkinsci/github-branch-source-plugin/pull/804/commits/d7a0bf399a732a417ba188df4b025bf382a120f5

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java (hunk 4 above)
Comment (jeromepochat) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java]: https://github.com/jenkinsci/github-branch-source-plugin/pull/804/commits/d7a0bf399a732a417ba188df4b025bf382a120f5

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java (hunk 3 above)
Comment (jeromepochat) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java]: https://github.com/jenkinsci/github-branch-source-plugin/pull/804/commits/d7a0bf399a732a417ba188df4b025bf382a120f5

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java (hunk 5 above)
Comment (jeromepochat) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GithubAppCredentialsContextualizationTest.java]: https://github.com/jenkinsci/github-branch-source-plugin/pull/804/commits/d7a0bf399a732a417ba188df4b025bf382a120f5

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubApp.java [hunk 7]
```diff
@@ -0,0 +1,51 @@
+package org.jenkinsci.plugins.github_branch_source;
+
+import com.cloudbees.plugins.credentials.CredentialsScope;
+import hudson.util.Secret;
+
+public class GitHubApp {
+
+    // PKCS8 private key (https://stackoverflow.com/a/22176759/4951015)
+    private static final String PRIVATE_KEY = "-----BEGIN PRIVATE KEY-----\n"
+            +
```
Comment (jglick) [inline src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubApp.java:10]: BTW when updating to 2.479.x we can switch to a multiline text block which would be more legible.

File: src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentialsContextualizationTest.java [hunk 8]
```diff
@@ -0,0 +1,95 @@
...
+                .willReturn(aResponse()
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBodyFile("../contextualization/body-repos-cloudbeers-multibranch-demo-contents.json")));
+        githubApi.stubFor(get(urlEqualTo("/repos/cloudbeers/multibranch-demo/pulls?state=open"))
+                .willReturn(aResponse()
+                        .withHeader("Content-Type", "application/json; charset=utf-8")
+                        .withBodyFile("../contextualization/body-repos-cloudbeers-multibranch-demo-pulls.json")));
+    }
+
+    @Test
```
[the remark this answers is not in the document]
Comment (jeromepochat) [reply src/test/java/org/jenkinsci/plugins/github_branch_source/GitHubAppCredentialsContextualizationTest.java:69]: https://github.com/jenkinsci/github-branch-source-plugin/pull/804/commits/24cb39b5ca680562081cf04c51cb63e4f7b446bc

**Pull request merged**